server/games/ladder_game.py

```diff
@@ -1,23 +1,61 @@
+import asyncio
 import logging
 from typing import Optional
 
 from server.config import config
 from server.players import Player
-from server.rating import RatingType
 
 from .game import Game
 from .game_results import ArmyOutcome, GameOutcome
-from .typedefs import FeaturedModType, GameType, InitMode
+from .typedefs import GameState, GameType, InitMode
 
 logger = logging.getLogger(__name__)
 
 
+class GameClosedError(Exception):
+    """
+    The game has been closed during the setup phase
+    """
+
+    def __init__(self, player: Player, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.player = player
+
+
 class LadderGame(Game):
     """Class for 1v1 ladder games"""
 
     init_mode = InitMode.AUTO_LOBBY
     game_type = GameType.MATCHMAKER
 
+    def __init__(self, id_, *args, **kwargs):
+        super().__init__(id_, *args, **kwargs)
+        self._launch_future = asyncio.Future()
+
+    async def wait_hosted(self, timeout: float):
+        return await asyncio.wait_for(
+            self._hosted_event.wait(),
```

Comment:
Another thing I just thought of. Maybe we want to apply the same cancellation logic for the host as well? Like if the host opens the game I to the Idle state but then closes it again before getting to the Lobby state.

Reply:
yes, good idea